engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java

```diff
@@ -111,29 +129,80 @@ long hashCodeForAtomConstructor(AtomConstructor atomConstructor) {
     return System.identityHashCode(atomConstructor);
   }
 
-  /** How many {@link HashCodeAnyNode} nodes should be created for fields in atoms. */
+  @Specialization
+  long hashCodeForUnresolvedSymbol(UnresolvedSymbol unresolvedSymbol,
+      @Cached HashCodeNode hashCodeNode) {
+    long nameHash = hashCodeNode.execute(unresolvedSymbol.getName());
+    long scopeHash = hashCodeNode.execute(unresolvedSymbol.getScope());
+    return Objects.hashCode(nameHash, scopeHash);
+  }
+
+  @Specialization
+  long hashCodeForUnresolvedConversion(UnresolvedConversion unresolvedConversion,
+      @CachedLibrary(limit = "5") InteropLibrary interop) {
+    return hashCodeForModuleScope(unresolvedConversion.getScope(), interop);
+  }
+
+  @Specialization
+  long hashCodeForModuleScope(ModuleScope moduleScope,
+      @CachedLibrary(limit = "5") InteropLibrary interop) {
+    return hashCodeForModule(moduleScope.getModule(), interop);
+  }
+
+  @Specialization
+  @TruffleBoundary
+  long hashCodeForModule(Module module,
+      @CachedLibrary(limit = "5") InteropLibrary interop) {
+    return hashCodeForString(module.toString(), interop);
+  }
+
+  @Specialization
+  long hashCodeForFile(EnsoFile file,
+      @CachedLibrary(limit = "5") InteropLibrary interop) {
+    return hashCodeForString(file.getPath(), interop);
+  }
+
+  /**
+   * There is no specialization for {@link TypesLibrary#hasType(Object)}, because also
+   * primitive values would fall into that specialization and it would be too complicated
+   * to make that specialization disjunctive. So we rather specialize directly for
+   * {@link Type}.
+   */
+  @Specialization
+  long hashCodeForType(Type type,
+      @Cached HashCodeNode hashCodeNode) {
+    if (EnsoContext.get(this).getNothing() == type) {
+      // Nothing should be equal to `null`
+      return 0;
```

Comment:
Does the hash of `null` have to be `0`? Probably not, but no problem making it so...

Reply:
It does not have to be 0, it could potentially be any value. But JDK returns 0 for nulls, so I just wanted to be consistent with JDK.